src/dsl.jl

```diff
@@ -298,149 +266,113 @@ end
-    forbidden_symbol_check(union(species, parameters))
-    forbidden_variable_check(variables)
-    unique_symbol_check(union(species, parameters, variables, ivs))
+    sps_inferred, ps_pre_inferred = extract_sps_and_ps(reactions, syms_declared; requiredec)
+    vs_inferred, diffs_inferred, equations = read_equations_options!(diffsexpr, options,
+        union(syms_declared, sps_inferred), tiv; requiredec)
+    ps_inferred = setdiff(ps_pre_inferred, vs_inferred, diffs_inferred)
+    syms_inferred = union(sps_inferred, ps_inferred, vs_inferred, diffs_inferred)
+
+    # Read options not related to the declaration or inference of symbols.
+    obsexpr, obs_eqs, obs_syms = read_observed_options(options, ivs,
+        union(sps_declared, vs_declared), union(syms_declared, syms_inferred); requiredec)
+    continuous_events_expr = read_events_option(options, :continuous_events)
+    discrete_events_expr = read_events_option(options, :discrete_events)
+    default_reaction_metadata = read_default_noise_scaling_option(options)
+    combinatoric_ratelaws = read_combinatoric_ratelaws_option(options)
 
     # Creates expressions corresponding to actual code from the internal DSL representation.
-    sexprs = get_sexpr(species_extracted, options; iv_symbols = ivs)
-    vexprs = get_sexpr(vars_extracted, options, :variables; iv_symbols = ivs)
-    pexprs = get_pexpr(parameters_extracted, options)
-    ps, pssym = assign_expr_to_var(!isempty(parameters), pexprs, "ps")
-    vars, varssym = assign_expr_to_var(!isempty(variables), vexprs, "vars";
-        scalarize = true)
-    sps, spssym = assign_expr_to_var(!isempty(species), sexprs, "specs"; scalarize = true)
-    comps, compssym = assign_expr_to_var(!isempty(compound_species), compound_expr,
-        "comps"; scalarize = true)
-    rxexprs = :(CatalystEqType[])
-    for reaction in reactions
-        push!(rxexprs.args, get_rxexprs(reaction))
-    end
-    for equation in equations
-        equation = escape_equation_RHS!(equation)
-        push!(rxexprs.args, equation)
-    end
-
-    # Output code corresponding to the reaction system.
-    quote
-        $ivexpr
-        $ps
-        $vars
-        $sps
-        $observed_vars
-        $comps
-        $diffexpr
-
-        sivs_vec = $sivs
-        rx_eq_vec = $rxexprs
-        vars = setdiff(union($spssym, $varssym, $compssym), $obs_syms)
-        obseqs = $observed_eqs
-        cevents = $continuous_events_expr
-        devents = $discrete_events_expr
+    psexpr_init = get_psexpr(ps_inferred, options)
+    spsexpr_init = get_usexpr(sps_inferred, options; ivs)
+    vsexpr_init = get_usexpr(vs_inferred, options, :variables; ivs)
+    psexpr, psvar = scalarize_macro(psexpr_init, "ps")
+    spsexpr, spsvar = scalarize_macro(spsexpr_init, "specs")
+    vsexpr, vsvar = scalarize_macro(vsexpr_init, "vars")
+    cmpsexpr, cmpsvar = scalarize_macro(cmpexpr_init, "comps")
+    rxsexprs = get_rxexprs(reactions, equations, union(diffs_declared, diffs_inferred))
```

Comment:
trying to harmonise notation a bit across everything. exprs is the expression that generates the actually code, var is the generated variable we store stuff in. init on things that should not be evaluated. sps are species, vs variables, us unknowns, ps parameters, cmps compounds. Syms are lists of symbols (e.g. the symbols corresponding to observables). Not insisting we use this across the board, just explaining how I have changed things.